Documentation/ApiOverview/RequestHandling/Index.rst

```diff
@@ -0,0 +1,271 @@
+
+* Create and return a PSR-7 response.
+
+* Call next request handler (which again will be a middleware).
+
+* Adjust response received from calling next request handler.
+
+.. _request-handling-middleware-examples:
+
+Middleware examples
+===================
+
+The following list shows typical use cases for middlewares.
+
+.. _request-handling-returning-custom-response:
+
+Returning custom response
+-------------------------
+
+This middleware will check whether TYPO3 is in maintenance mode and will return
+an unavailable response in that case. Otherwise the next middleware will be
+called, and their response is returned instead.
+
+::
+
+   public function process(
+       ServerRequestInterface $request,
+       RequestHandlerInterface $handler
+   ): ResponseInterface {
+       if (/* if logic */) {
+           return GeneralUtility::makeInstance(ErrorController::class)
+               ->unavailableAction(
+                   $request,
+                   'This page is temporarily unavailable.'
+               );
+       }
+
+       return $handler->handle($request);
+   }
+
+.. _request-handling-enriching-request:
+
+Enriching request
+-----------------
+
+The current request can be extended with further information, e.g. the current
+resolved site and language could be attached to the request.
+
+In order to do so, a new request is built with additional attributes, before
+calling the next request handler with enhanced request.
+
+::
+
+   public function process(
+       ServerRequestInterface $request,
+       RequestHandlerInterface $handler
+   ): ResponseInterface {
+       $routeResult = $this->matcher->matchRequest($request);
+
+       $request = $request->withAttribute('site', $routeResult->getSite());
+       $request = $request->withAttribute('language', $routeResult->getLanguage());
+
+       return $handler->handle($request);
+   }
+
+.. _request-handling-enriching-response:
+
+Enriching response
+-----------------
+
+This middleware will check the length of generated output, and add an header
+with this information to the response.
+
+In order to do so, next request handler is called. He will return the generated
+response, which can be enriched before it gets returned.
+
+::
+
+   public function process(
+       ServerRequestInterface $request,
+       RequestHandlerInterface $handler
+   ): ResponseInterface {
+       $response = $handler->handle($request);
+
+       if (/* if logic */) {
+           $response = $response->withHeader(
+               'Content-Length',
+               (string)$response->getBody()->getSize()
+           );
+       }
+
+       return $response;
+   }
+
+.. _request-handling-configuring-middlewares:
+
+Configuring middlewares
+=======================
+
+In order to implement a custom middleware, this middleware has to be configured.
+TYPO3 already provides some middlewares pre configured. Beside adding own
+middlewares, it's also possible to remove existing middlewares from
+configuration.
```

Comment:
"the configuration" or introduce the term "middleware stack" in the beginning to use it here.